rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+  proto.MFAAuthenticateResponse mfa_response = 3;
+}
+```
+
+#### `rpc GenerateUserCerts`
+
+This is an existing endpoint used to generate user certs. We will add the `headless_request_id` and `request_token` to retrieve request details from the headless auth request saved on the Auth server.
+
+### UX
+
+#### `tsh --headless`
+
+We will add a new `--headless` flag to `tsh` which can be used to authenticate for a single `tsh` request. When this flag is provided, `tsh` will prompt the user to complete headless authentication on their local machine from the URL `https://proxy.example.com/headless/<request_id>/approve`. Once the user completes local authentication, `tsh` will receive credentials to complete the request.
+
+```console
+$ tsh --headless --proxy=proxy --user=user ssh user@node01
+Complete headless authentication in your local web browser:
+https://proxy.example.com/headless/<request_id>/approve
+// Wait for user to complete local authentication with MFA
+<user@node01> $
+```
+
+#### Environment variables
+
+In the `tsh --headless` flow, users never run `tsh login` on their remote machine. Instead, we expect the `--proxy`, `--user`, and `--headless` flags to be supplied to each command. To reduce UX friction, users can set the environment variables `TELEPORT_PROXY=<proxy_addr>`, `TELEPORT_USER=<user>`, and `TELEPORT_HEADLESS=true` instead.
+
+We prefer setting environment variables rather than saving config to disk (`~/.tsh/proxy.example.com.yaml` and `~/.tsh/current-profile`) so that the headless flow remains stateless, preventing conflicts on shared machines.
+
+#### `tsh headless`
+
+When the user goes to the headless authentication URL, the user will be prompted to login with MFA, if they are not logged in already. The user will then be notified of additional request details. Finally, the user is asked to verify with MFA to approve the request.
+
+Example (Exact UI/UX TBD):
+```
+Headless request requires authentication. Contact your administrator if you didn't initiate this request.
+Additional details:
+  - command: "tsh ssh user@localhost"
+  - request id: <request_id>
+  - ip address: <ip_address>
+Tap your YubiKey to approve
+```
+
+Note: When the user has to log in for the first time, we do not reuse their MFA verification to skip the second MFA check. Although this would be better UX, we cannot retrieve additional request details to share with the user until they log in. For security reasons, we should provide an MFA check after sharing the headless request details.
+
+#### View headless requests
+
+In the Web UI, we will create a new page to view and accept `tsh` requests for a headless session: `https://proxy.example.com/headless/<headless_session_id>/requests`. The UI may be very similar to the access request page, where a user can view requests, view additional details, and then click approve/deny. When the user clicks "approve", this will trigger a prompt for MFA verification to complete the approval.
+
```

Comment:
What other alternatives did you consider? We've had various features requests for this kind of functionality over time, with various suggestions on how to solve it as well. It would be interesting to mention those and say why we chose this approach instead.

Reply:
Alternatives to what in particular?

Reply:
I realize it's a bit vague, but to the design itself. Did you consider other approaches? How did they compare?

For example, we could try and pipe MFA challenges back through the SSH tunnel, use Yubico OTP, etc - all of these popped into feature requests one time or another.

Reply:
Given the timeline, I mostly stuck to this single design approach. The general idea that I built around was to have an SSO flow that could authenticate a user from a remote login session with just the Teleport Auth/Proxy, with MFA check added on.

Reply:
Fair enough, thanks.